shared/utility/math/stats/multivariate.hpp

```diff
@@ -0,0 +1,119 @@
+/*
+ * This file is part of the NUbots Codebase.
+ *
+ * The NUbots Codebase is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The NUbots Codebase is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright 2013 NUbots <nubots@nubots.net>
+ */
+
+#ifndef UTILITY_MATH_STATS_MULTIVARIATE_HPP
+#define UTILITY_MATH_STATS_MULTIVARIATE_HPP
+
+#include <Eigen/Core>
+#include <Eigen/Eigenvalues>
+#include <cmath>
+#include <random>
+
+namespace utility::math::stats {
+
+    /**
+     * @author Alex Biddulph
+     *
+     * @brief N dimension multivariate normal distribution
+     * Implementation taken from https://stackoverflow.com/a/40245513/4795763 and
+     * https://en.wikipedia.org/wiki/Multivariate_normal_distribution
+     *
+     * @tparam Scalar The scalar type to use for all calculations.
+     * @tparam N The dimensionality of the multivariate distribution
+     */
+    template <typename Scalar, int N>
+    struct MultivariateNormal {
+    private:
+        // Must construct with at least a covariance matrix
+        MultivariateNormal() = delete;
+
+    public:
+        /**
+         * @brief Construct a multivariate normal distribution of dimension N with zero mean and the provided covariance
+         *
+         * @param covariance_ The covariance matrix for this zero-mean multivariate normal distribution
+         */
+        MultivariateNormal(const Eigen::Matrix<Scalar, N, N>& covariance_)
+            : MultivariateNormal(Eigen::Matrix<Scalar, N, 1>::Zero(), covariance_) {}
+
+        /**
+         * @brief Construct a multivariate normal distribution of dimension N with the provided mean and covariance
+         *
+         * @param mean_ The mean for this mean multivariate normal distribution
+         * @param covariance_ The covariance matrix for this multivariate normal distribution
+         */
+        MultivariateNormal(const Eigen::Matrix<Scalar, N, 1>& mean_, const Eigen::Matrix<Scalar, N, N>& covariance_)
+            : mean(mean_), covariance(covariance_), gen(rd()) {
```

Comment:
No need to store a random device we just need to use it once to seed the pseudorandom number generator
```suggestion
            : mean(mean_), covariance(covariance_), gen(std::random_device()()) {
```